driver-sync/src/test/functional/com/mongodb/client/ClientSideEncryptionCorpusTest.java

```diff
@@ -0,0 +1,292 @@
+            if (field.equals("_id") || field.equals("altname_aws") || field.equals("altname_local")) {
+                continue;
+            }
+
+            String kms = corpusEncryptedActual.getDocument(field).getString("kms").getValue();
+            String type = corpusEncryptedActual.getDocument(field).getString("type").getValue();
+            String algo = corpusEncryptedActual.getDocument(field).getString("algo").getValue();
+            String method = corpusEncryptedActual.getDocument(field).getString("method").getValue();
+            // TODO: unused
+            String identifier = corpusEncryptedActual.getDocument(field).getString("identifier").getValue();
+            boolean allowed = corpusEncryptedActual.getDocument(field).getBoolean("allowed").getValue();
+            BsonValue value = corpusEncryptedActual.getDocument(field).get("value");
+
+            // All deterministic fields are an exact match.
+            if (algo.equals("det")) {
+                deterministicCheck(kms, type, corpusEncryptedExpected);
+            }
+
+            if (algo.equals("rand") && !method.equals("prohibited")) {
+                randomCheck(value, field, corpusEncryptedExpected);
+            }
+
+            if (allowed) {
+                BsonValue decrypted = clientEncryption.decrypt(value.asBinary());
+                assertEquals("Values should be equal for field " + field, corpus.getDocument(field).get("value"), decrypted);
+            } else {
+                assertEquals("Values should be equal for field " + field, corpus.getDocument(field).get("value"),
+                        corpusEncryptedActual.getDocument(field).get("value"));
+            }
+        }
+    }
+
+    // Check that all values of document that are deterministic with the same kms + type match value
+    private static void deterministicCheck(final String matchKms, final String matchType, final BsonDocument document) {
+        for (String field : document.keySet()) {
+            if (field.equals("_id") || field.equals("altname_aws") || field.equals("altname_local")) {
+                continue;
+            }
+
+            BsonDocument subDocument = document.getDocument(field);
+
+            String kms = subDocument.getString("kms").getValue();
+            String type = subDocument.getString("type").getValue();
+            String algo = subDocument.getString("algo").getValue();
+            BsonValue value = subDocument.get("value");
+
+            if (kms.equals(matchKms) && type.equals(matchType) && algo.equals("det")) {
+                assertEquals(value, value);
+            }
+        }
+    }
+
+    private static void randomCheck(final BsonValue value, final String excludeField, final BsonDocument document) {
+        for (String field : document.keySet()) {
+            if (field.equals(excludeField)) {
+                continue;
+            }
+            assertNotEquals(document.get(field), value);
```

Comment:
This is comparing the encrypted bindata with the full field's document `{ "value": "...", "type": "...", "algo": "..." }`, so it's never equal – but the check is wrong. Testing with my local runner, I think it should be updated to the following:

```
    private static void randomCheck (BsonValue val, String exclude_field, BsonDocument doc) {
        for (String field : doc.keySet()) {
            if (field.equals("_id") || field.equals("altname_aws") || field.equals("altname_local")) {
                continue;
            }

            if (field.equals(exclude_field)) {
                continue;
            }

            assert !doc.getDocument(field).get("value").equals(val);
        }
    }
```

I verified that this check fails when comparing a disallowed field, and succeeds with the fix above to checking `allowed` instead of `prohibited`.

Reply:
Done